File: test/unit/markdownParser.test.js
// @ts-check

import assert from "node:assert/strict";
import { test } from "node:test";

import { parseMarkdownHeading } from "../../src/parsers/markdownParser.js";

test("em-dash heading parses", () => {
  const parsed = parseMarkdownHeading("## src/util/foo.js — checkout-flow");

  assert.deepEqual(parsed, {
    source: "src/util/foo.js",
    anchor: "checkout-flow",
    start: "## ".length,
    end: "## ".length + "src/util/foo.js".length,
  });
});

test("space-hyphen heading parses", () => {
  const parsed = parseMarkdownHeading("## src/util/foo.js - checkout-flow");

  assert.equal(parsed.source, "src/util/foo.js");
  assert.equal(parsed.anchor, "checkout-flow");
});

test("hash-separated heading parses (legacy tolerance)", () => {
  const parsed = parseMarkdownHeading("## src/checkout/cart.js" + "#checkout-flow");

  assert.equal(parsed.source, "src/checkout/cart.js");
  assert.equal(parsed.anchor, "checkout-flow");
});

test("heading without anchor does not parse", () => {
  assert.equal(parseMarkdownHeading("## src/util/foo.js"), null);
});

test("single-hash heading does not parse", () => {
  assert.equal(parseMarkdownHeading("# src/util/foo.js — anchor"), null);
});

test("non-heading line does not parse", () => {
  assert.equal(parseMarkdownHeading("some prose text here"), null);
});

File: jetbrains/src/test/kotlin/com/anchor/commentdoclinks/resolver/MarkdownParserTest.kt
package com.anchor.commentdoclinks.resolver

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class MarkdownParserTest {
    @Test
    fun `parses em dash heading`() {
        val h = parseMarkdownHeading("## src/checkout/cart.js — checkout-flow")!!
        assertEquals("src/checkout/cart.js", h.source)
        assertEquals("checkout-flow", h.anchor)
    }

    @Test
    fun `parses dash heading`() {
        val h = parseMarkdownHeading("## src/checkout/cart.js - checkout-flow")!!
        assertEquals("src/checkout/cart.js", h.source)
        assertEquals("checkout-flow", h.anchor)
    }

    @Test
    fun `parses hash heading`() {
        val h = parseMarkdownHeading("## src/checkout/cart.js#checkout-flow")!!
        assertEquals("src/checkout/cart.js", h.source)
        assertEquals("checkout-flow", h.anchor)
    }

    @Test
    fun `source span excludes separator and anchor`() {
        val line = "## src/checkout/cart.js — checkout-flow"
        val h = parseMarkdownHeading(line)!!
        assertEquals(line.indexOf("src/checkout/cart.js"), h.start)
        assertEquals(line.indexOf("src/checkout/cart.js") + "src/checkout/cart.js".length, h.end)
    }

    @Test
    fun `rejects level one heading`() {
        assertNull(parseMarkdownHeading("# src/checkout/cart.js — checkout-flow"))
    }

    @Test
    fun `rejects heading without anchor separator`() {
        assertNull(parseMarkdownHeading("## just a normal heading"))
    }
}
